profiling/src/php_ffi.c

```diff
@@ -142,7 +156,83 @@ void ddog_php_prof_zend_mm_set_custom_handlers(zend_mm_heap *heap,
 #endif
 }
 
-zend_execute_data* ddog_php_prof_get_current_execute_data()
-{
+zend_execute_data* ddog_php_prof_get_current_execute_data() {
     return EG(current_execute_data);
 }
+
+#if CFG_RUN_TIME_CACHE // defined by build.rs
+static int ddog_php_prof_run_time_cache_handle = -1;
+#endif
+
+void ddog_php_prof_function_run_time_cache_init(const char *module_name) {
+#if CFG_RUN_TIME_CACHE // defined by build.rs
+    // Grab 2, one for function name and one for filename.
+#if PHP_VERSION_ID < 80200
+    ddog_php_prof_run_time_cache_handle =
+        zend_get_op_array_extension_handle(module_name);
+    int second = zend_get_op_array_extension_handle(module_name);
+    ZEND_ASSERT(ddog_php_prof_run_time_cache_handle + 1 == second);
+#else
+    ddog_php_prof_run_time_cache_handle =
+        zend_get_op_array_extension_handles(module_name, 2);
+#endif
+#else
+    (void)module_name;
+#endif
+
+    /* It's possible to work on PHP 7.4 as well, but there are opcache bugs
+     * that weren't truly fixed until PHP 8:
+     * https://github.com/php/php-src/pull/5871
+     * I would rather avoid these bugs for now.
+     */
+}
+
+#if CFG_RUN_TIME_CACHE // defined by build.rs
+static bool has_invalid_run_time_cache(zend_function const *func) {
+    if (_ignore_run_time_cache) return true;
+
+    // It should be initialized by this point, or we failed.
+    bool is_not_initialized = ddog_php_prof_run_time_cache_handle < 0;
+
+    // Trampolines use the extension slot for internal things.
+    bool is_trampoline = func->common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE;
+
+#if PHP_VERSION_ID < 80200
+    // Internal functions don't have a runtime cache until PHP 8.2.
+    bool is_internal = func->type == ZEND_INTERNAL_FUNCTION;
+
+    // In some cases, using the logical-or instead of bitwise-or will end up
+    // having conditional jumps. Since we overwhelmingly expect all conditions
+    // to be false, reducing the branching helps a tiny bit for performance.
+    return is_not_initialized | is_trampoline | is_internal;
+#else
+    return is_not_initialized | is_trampoline;
+#endif
+}
+#endif
+
+uintptr_t *ddog_php_prof_function_run_time_cache(zend_function const *func) {
+#if CFG_RUN_TIME_CACHE
+    if (UNEXPECTED(has_invalid_run_time_cache(func))) return NULL;
+
+#if PHP_VERSION_ID < 80200
+    // Internal functions don't have a runtime cache until PHP 8.2.
+    uintptr_t *cache_addr = RUN_TIME_CACHE(&func->op_array);
+#else
+    uintptr_t *cache_addr = RUN_TIME_CACHE(&func->common);
+#endif
+
+    // todo: how sure can I be, here? Can I be so confident as to omit it?
+    ZEND_ASSERT(cache_addr);
```

Comment:
@bwoebi Can you think of any reason for the `cache_addr` to be null, given the guards in `has_invalid_run_time_cache`?

Reply:
No. That's it.

Reply:
Thanks. I need to update the comment then.